### ISSUE TITLE
Infinite RAM consumption over time on Firefox

### ISSUE THREAD

**Issue by pahofmann** [mainsail-crew/mainsail#904]
### Browser:

Firefox

### Device:

Desktop PC

### Operating System:

Linux

### What happened?

If mainsail has a webcam configured and mainsail is left open in firefox it keeps consuming more and more RAM till the system runs out of memory/swap.

This doesn't happen on chrome.

### What did you expect to happen instead?

Normal RAM usage.

### How to reproduce this bug?

Open Mainsail in Firefox on a printer with a webcam configures. 
Leave it running for a few hours.

**Comment by th33xitus** [comment]
Which Mainsail version do you use?
With the release of 2.2.0 we fixed a few webcam related issues. Including a memory leak.

**Comment by Joel-De** [comment]
This seems to still be an issue on 2.10 (updated to 2.11 last night), though it seems to be an edge case since it works fine most of the time. That said on 2 separate instances mainsail (on firefox) has consumed 7GB of ram and crashed the system. I am also using Crowsnest and have timelapses set up so there is that commonality. Don't have logs at this time but I'll create a new issue if it happens again with more details.

**Comment by KungfuPancake** [comment]
I have the same issue on two separate computers, both Firefox, one on Archlinux and one on Windows 10. The problem is very likely coupled to the Webcam, as the problem is exacerbated by opening the large "Webcam" view.

**Comment by Joel-De** [comment]
> I have the same issue on two separate computers, both Firefox, one on Archlinux and one on Windows 10. The problem is very likely coupled to the Webcam, as the problem is exacerbated by opening the large "Webcam" view.

@KungfuPancake I'm out of the city and won't have access to logs for about a week, if you're able to upload them and create an issue feel free to do that and link it here. I tried reproducing the issue today but wasn't able to get FireFox to use 100% of my RAM though it's entirely possible that I didn't wait long enough as I only had an hour.

**Comment by meteyou** [comment]
@KungfuPancake @Joel-De which webcam modes do you use?

**Comment by Joel-De** [comment]
@meteyou uStreamer & hyperlapse with the interval set to ~5s. I'm using 720p as the camera resolution so nothing too high.

**Comment by meteyou** [comment]
@Joel-De which mode do you use in mainsail? The streamer itself doesn't matter...

**Comment by Joel-De** [comment]
@meteyou I'm using local mode to host mainsail, used KIAUH to install the package and the rest of Klipper. I'm only using a single printer in my configuration.

**Comment by meteyou** [comment]
@Joel-De i mean your webcam settings in mainsail itself. There are multiple types you could run your webcam in Mainsail.

**Comment by meteyou** [comment]
@Joel-De no... i dont need your timelapse settings. timelapse will only be executing on the host (pi) itself. i need your webcam setting in mainsail! (no timelapse, no crowsnest...)

**Comment by Joel-De** [comment]
@meteyou Are you referring to the items [here](https://docs.mainsail.xyz/overview/settings/webcams )? If not could you give me a link to the settings you're talking about? Thanks

**Comment by meteyou** [comment]
@Joel-De yes, this page contains all webcam settings in Mainsail. Pls post a screenshot from each webcam.

**Comment by Joel-De** [comment]
Here's a screenshot, I only have a single webcam setup on my system.

![image](https://github.com/mainsail-crew/mainsail/assets/71618484/51421861-a438-4e93-ad61-e0dbf0518fb3)

**Comment by meteyou** [comment]
ok. can you double check, if you also get this error without a webcam in mainsail? pls don't forget to refresh 1 time after removing the webcam in this settings.

**Comment by Joel-De** [comment]
I haven't been able to reproduce this issue with the camera disabled. That said I'm also not able to reliably reproduce it with it on either, although I did have this issue pop up yesterday after what I would estimate to be ~6 hours of runtime.

It doesn't seem to be slow building either - everything will be fine for several hours and it will abruptly balloon. I'll monitor it closely today and see if I can provide more hints. Are there any logs that would be useful for you to debug this?

**Comment by meteyou** [comment]
Not really... This is the problem with this issue...

**Comment by Joel-De** [comment]
Alright, tried it out on my desktop since I'm able to interact with the laptop that normally runs mainsail directly. Reproduced the bug though again it seems to appear randomly.

Memory snapshots of Mainsail didn't show any growth (~90MB of usage measured via Inspect -> Memory on firefox), in the meantime the main firefox process was growing pretty rapidly (~1GB -> ~2GB). From what firefox provides, "heap-unclassified" and "js-non-window" seem to be the areas in the main process consuming the most ram.

After about 7 hours my laptop that runs klipper and mainsail OOM'd and crashed. Didn't interact with it after bootup and it was on a screensaver after the first 30 minutes.

**Comment by meteyou** [comment]
Normally, the browser go into sleep mode and mainsail should be on "pause"

**Comment by Joel-De** [comment]
Yeah that's why I was surprised when it OOM'd while the screen had been off for several hours, I didn't think that was possible. I'll be switching from firefox to a chromium browser / kiosk viewer since some of these crashes have resulted in print failures, so I probably won't have too much more information to share here.

Feel free to re-open this issue or to create a new one if you feel that's justified.

Thanks

**Comment by TECH-GEEK** [comment]
This bug is present in v2.12.0. I have one tab open with mainsail and it goes up to 16gb of ram usage for one tab in firefox in less than an hour.

**Comment by meteyou** [comment]
@TECH-GEEK on how many places do you want to "spam" this now?

**Comment by TECH-GEEK** [comment]
Sorry. I have autism and poor understanding of social skills.  I didn't mean to spam. Feel free to delete things

**Comment by meteyou** [comment]
@TECH-GEEK ah ok. Np. Plus open the post in the discord support forum. I hope we can reproduce this issue, that we can fix it.

**Comment by Monstrofil** [comment]
<img width="1919" height="1039" alt="Image" src="https://github.com/user-attachments/assets/42fc8afe-2d50-48e8-9e1c-836140ea4156" />

Anyone knows if this issue was ever fixed?

<img width="348" height="111" alt="Image" src="https://github.com/user-attachments/assets/5920b64c-3ae1-4c4f-88e3-94fdeda19b28" />

**Comment by rackrick** [comment]
@Monstrofil this was fixed with https://github.com/mainsail-crew/mainsail/pull/1987
